In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

1. Import the datasets

In [7]:
# Set path
path = r'/Users/christina/Desktop/Instcart Basket Analysis'

In [9]:
# Import dataset orders_products_prior.csv
df_ords_prior = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [11]:
# Import dataset orders_cleaned.csv
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_cleaned.csv'), index_col = False)

2. Check the dimension of the imported dataframes

In [13]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [15]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

In [17]:
df_ords_prior.shape


(32434489, 4)

In [19]:
df_ords.shape

(3421083, 9)

3. Merge the dataframes

In [22]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [23]:
df_merged_large.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

4. Use the indicator argument to check whether there was a full match between the two dataframes

In [33]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

5. Export data in pickle format

In [24]:
# Export data to pkl
df_merged_large.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_combined.pkl'))

In [25]:
# Export modified df_ords dataframe
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_final.csv'))

6. import datasets

In [26]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [27]:
# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

7. Check the dimensions of the imported dataframes

In [28]:
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [43]:
df_ords_prods_combined.shape

(32434489, 13)

In [31]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [32]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [33]:
df_prods.shape

(49672, 6)

8. Determine a suitable way to combine the orders_products_combined dataframe with the products data set

In [53]:
# A suitable way to combine the df_ords_prods_combined dataframe with the df_prods dataframe, despite their different shape, 
# is to merge them thanks to their shared column "product_id". 
# For the purpose of this project, a default inner join is advisable

In [34]:
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = '_merge2')

In [36]:
# Check output
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [37]:
df_ords_prods_combined.shape

(32434489, 13)

In [49]:
# Merge the updated dataframes
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = '_merge2')


In [50]:
# Check the output
df_ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id eval_set  \
0             19     5.8          1987   3139998      138    prior   
1             19     5.8          1989   1977647      138    prior   
2             19     5.8         11433    389851      709    prior   
3             19     5.8         12198    652770      764    prior   
4             19     5.8         12200   1813452      764    prior   

   order_number  orders_day_of_week  order_hour_of_day  \
0            28                   6                 11   
1            30                   6                 17   
2             2                   0                 21   
3             1                   3                 13   
4             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

  _merge2  
0    both  
1    both  
2    both  
3    both  
4    both

In [51]:
df_ords_prods_merge.shape

(32404859, 19)

8. Confirm the results of the merge using the merge flag

In [52]:
df_ords_prods_merge['_merge2'].value_counts()

_merge2
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project.

9. Export the newly created dataframe as ords_prods_merge in a suitable format

In [53]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'ords_prods_merge.pkl'))